src/Functions/countDigits.cpp

```diff
@@ -92,50 +93,22 @@ class FunctionCountDigits : public IFunction
         for (size_t i = 0; i < rows_count; ++i)
         {
             if constexpr (is_decimal<T>)
-                dst_data[i] = digits<NativeT>(src_data[i].value);
+            {
+                auto value = src_data[i].value;
+                if (unlikely(value < 0))
+                    dst_data[i] = digits10<NativeT>(-static_cast<NativeT>(value));
```

Comment:
FYI first casting to unsigned, then negating - is the only safe way to do it in C++.